Add KISTERS case study (#2637)
mainmatter/mainmatter.com: src/cases/kisters.md

---
layout: case-study
company: KISTERS

> problem: KISTERS wanted to use Rust's cross-platform support to its full potential.

Reviewer: This doesn't say much IMO, plus it sounds like they wanted an excuse to use Rust rather than wanted to solve a problem (or maybe I misunderstood?)

Author: ```suggestion
problem: KISTERS wanted to dynamically move computation to different platforms to handle traffic spikes.
```

@KevinBongart How about this?

Reviewer: That's better. Going even further, they needed to handle traffic spikes by dynamically moving computation to different platforms (but I'm nitpicking)

> solution: We created a data processing package suitable for server, edge, and WASM.

Reviewer: ```suggestion
solution: We created a data-processing package suitable for server, edge, and WASM.
```

description: <p>KISTERS builds solutions for collecting, analyzing, and providing information based on environmental data. The HydroMet division provides software that helps warn people of upcoming floods using data obtained from numerous sensor devices deployed across the globe.</p>
hero:
  color: purple
  image: "/assets/images/work/aleph-alpha-background-2.jpg" # TODO
  imageAlt: "TODO" # TODO
  tags: "development / architecture / Rust"
og:
  image: /assets/images/cases/cs-aleph-alpha-og-image.jpg # TODO
---

{% from "image-aspect-ratio.njk" import imageAspectRatio %} {% from "quote.njk" import quote %}

<div class="case-study__section">
  <h3 class="case-study__heading">About KISTERS</h3>
  <div class="case-study__text">
    <p><a href="http://kisters.eu/">KISTERS</a> provides software for collecting and analyzing environmental data.<br>
      In particular, they provide solutions to warn people of upcoming floods using data obtained from numerous sensor devices deployed throughout the globe.</p>
  </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">The challenge</h3>
    <div class="case-study__text">

>       <p>To be able to <strong>handle large traffic spikes efficiently without risking astronomic cloud expenses, KISTERS needed the ability to dynamically move computation</strong> between traditional servers and edge functions, or offloading to end user's devices.</p>

Reviewer 2: I would break this down into two sentences, a flip it around (first what they need to do, then why).

>       <p>To be able to <strong>handle large traffic spikes efficiently without risking astronomic cloud expenses, KISTERS needed the ability to dynamically move computation</strong> between traditional servers and edge functions, or offloading to end user's devices.</p>

Reviewer: ```suggestion
      <p>To be able to <strong>handle large traffic spikes efficiently without risking astronomic cloud expenses, KISTERS needed the ability to dynamically move computation</strong> between traditional servers and edge functions, or offloading to end-user devices.</p>
```

>       <p>The large amounts of data KISTERS obtains from sensor devices needs processing to become useful for analysis. Filtering, grouping, and post-processing, depend on dynamic inputs and need to be carried out on demand.</p>

Reviewer: ```suggestion
      <p>The large amounts of data KISTERS obtains from sensor devices need processing to become useful for analysis. Filtering, grouping, and post-processing, depend on dynamic inputs and need to be carried out on demand.</p>
```

(large amounts is plural)

      <p>That in and of itself warrants a well-optimized design and use of performant technologies, but in times of calamity, <strong>the amount of users requesting information may rise quickly</strong>: residents of endangered areas can use KISTERS’ software to obtain real-time information regarding their situation.</p>
    </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">How Mainmatter helped out</h3>
    <div class="case-study__text">

>         <p>During an initial development cycle, Mainmatter proposed and implemented an architecture that isolates the data processing implementation in a platform-independent package written in Rust.<p>

Reviewer 2: ```suggestion
        <p>Mainmatter proposed and implemented an architecture that isolates the data processing implementation in a platform-independent package written in Rust.<p>
```

</div>

<section class="mt-5">
        <!-- TODO add image for KISTERS -->
        {% set imageData = {
          "imgPath": "/assets/images/work/aleph-alpha-graphs.jpg",
          "alt": "TODO",
          "sizes": "100vw",
          "loading": "lazy",
          "sizesArray": [760, 1440, 1920]
        } %}
        {{ imageAspectRatio(imageData, "32/13", "35/19") }}
</section>

<div class="case-study__section">
  <h3 class="case-study__heading">Technology</h3>
    <div class="case-study__text">
      <p></p>
      <p>Rust, with its strong performance and stability characteristics and solid WASM support, was a natural choice as a language with which to implement the design. What’s more, <a href="https://pola.rs/">Polars</a>, the famous high-performance open-source data frame manipulation library, is implemented in Rust itself and easily usable in Rust projects including those targeting WASM.</p>
    </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">Aftercare</h3>
    <div class="case-study__text">
      <p>Upon finishing and delivering the proof-of-concept, Mainmatter provided guidance to KISTERS' team on integrating the code into their platform.</p>

>       <p>Mainmatter has reviewed and optimized the integration code during several pairing sessions, ensuring that not just the product but also the knowledge required to maintain and expand it is transferred to KISTERS’ engineering team.</p>

Reviewer: ```suggestion
      <p>Mainmatter has reviewed and optimized the integration code during several pairing sessions with KISTERS’ engineering team, ensuring the delivery of both the product and the technical knowledge required to maintain and expand it.</p>
```

---
layout: case-study
company: KISTERS
problem: KISTERS wanted to use Rust's cross-platform support to its full potential.
solution: We created a data processing package suitable for server, edge, and WASM.
tags: Launch your idea
title: Dynamically switching execution platforms with Rust | Work
displayTitle: "Dynamically switching execution platforms with Rust"

> description: <p>KISTERS builds solutions for collecting, analyzing, and providing information based on environmental data.</p><p>The department we’ve been working with provides software that can help warn people of upcoming floods using data obtained from numerous sensor devices deployed throughout the globe.</p>

Reviewer 3: ```suggestion
description: <p>KISTERS builds solutions for collecting, analyzing, and providing information based on environmental data. The HydroMet division provides software that helps warn people of upcoming floods using data obtained from numerous sensor devices deployed across the globe.</p>
```

hero:
  color: purple
  image: "/assets/images/work/aleph-alpha-background-2.jpg" # TODO
  imageAlt: "TODO" # TODO
  tags: "development / architecture / Rust"
og:
  image: /assets/images/cases/cs-aleph-alpha-og-image.jpg # TODO
---

{% from "image-aspect-ratio.njk" import imageAspectRatio %} {% from "quote.njk" import quote %}

<div class="case-study__section">
  <h3 class="case-study__heading">About KISTERS</h3>
  <div class="case-study__text">

>     <p>Based in Aachen, Germany, <a href="http://kisters.eu/">KISTERS</a> provides software for collecting, analyzing, and providing information based on environmental data.<br>

Reviewer 2: I wouldn't lead with "Based in Aachen, Germany". It's not that relevant. If we want to include it, it should be later in the sentence or in a follow-up paragraph.

>       The department we’ve been working with provides solutions that can help warn people of upcoming floods using data obtained from numerous sensor devices deployed throughout the globe.</p>

Reviewer 2: ```suggestion
      In particular, they provide solutions to warn people of upcoming floods using data from numerous sensor devices deployed throughout the globe.</p>
```

  </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">The challenge</h3>
    <div class="case-study__text">

>       <p>KISTERS wanted to <strong>handle large traffic spikes by dynamially switching</strong> between serving from a traditional server, from an edge function, or by offloading cumputation to end user's devices.</p>

Reviewer 3: ```suggestion
      <p>To be able to <strong>handle large traffic spikes efficiently without risking astronomic cloud expenses, KISTERS needed the ability to dynamically move computation</strong> between traditional servers and edge functions, or offloading to end user's devices.</p>
```

>       <p>With big data comes big computation. KISTERS wanted to <strong>handle large traffic spikes by dynamially switching</strong> between serving from a traditional server, from an edge function, or by offloading cumputation to end user's devices.</p>

Reviewer 2: I don't think this is accurate. The data we're working with here is far from being "big data" (i.e. data that doesn't fit into the memory of a single machine).
WASM is not an option for big data, since you woulnd't be able to stream it/handle it on the client device.

The challenge is here is strictly related to the unpredictable and very spiky traffic shape.

>       <p>Why? Well, the data KISTERS obtains from their sensor devices needs processing to become useful for analysis. The concrete processing steps taken, such as filtering, grouping, and post-processing, depend on dynamic inputs reflecting the needs of KISTERS’ clients. Therefore, they need to be carried out on user demand.</p>
>       <p>That in and of itself warrants a well-optimized design and use of performant technologies, but in times of calamity, <strong>the amount of users requesting information may rise quickly</strong>: residents of endangered areas can use KISTERS’ software to obtain real-time information regarding their situation.</p>
>       <p><strong>To handle large demand efficiently</strong>, without risking astronomic cloud expenses, KISTERS’ sought to employ a design that would allow them to <strong>quickly deploy their data processing to edge functions</strong> and, in case of extreme traffic surges, allow the processing steps to be <strong>run within the browser</strong> on end user’s devices.</p>

Reviewer 3: ```suggestion
      <p>The large amounts os data KISTERS obtains from sensor devices needs processing to become useful for analysis. Filtering, grouping, and post-processing, depend on dynamic inputs and need to be carried out on demand.</p>
      <p>That in and of itself warrants a well-optimized design and use of performant technologies, but in times of calamity, <strong>the amount of users requesting information may rise quickly</strong>: residents of endangered areas can use KISTERS’ software to obtain real-time information regarding their situation.</p>
```

I'd move this above the paragraph above as it's kind of the ramp-up to what we're saying there

    </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">How Mainmatter helped out</h3>
    <div class="case-study__text">

>         <p>During an initial development cycle, Mainmatter proposed and implemented an architecture that isolates the implementation of the data processing steps in a platform-independent package.<p>

Reviewer 3: ```suggestion
        <p>During an initial development cycle, Mainmatter proposed and implemented an architecture that isolates the data processing implementation in a platform-independent package written in Rust.<p>
```

>         <p>That way, the package can be thinly wrapped for use by a web server, served as an edge function, or compiled to Web Assembly for execution in browsers, guaranteeing application behavior on different platforms to be equal.</p>

Reviewer 3: ```suggestion
        <p>That package can be thinly wrapped for use in a web server, run as an edge function, or compiled to Web Assembly (WASM) for execution in browsers. All of these different environments run the same efficient code with close to no duplication.</p>
```

        <p>That way, the package can be thinly wrapped for use by a web server, served as an edge function, or compiled to Web Assembly for execution in browsers.</p>
    </div>
</div>

<section class="mt-5">
        <!-->TODO add image for KISTERS<-->
        {% set imageData = {
          "imgPath": "/assets/images/work/aleph-alpha-graphs.jpg",
          "alt": "TODO",
          "sizes": "100vw",
          "loading": "lazy",
          "sizesArray": [760, 1440, 1920]
        } %}
        {{ imageAspectRatio(imageData, "32/13", "35/19") }}
</section>

<div class="case-study__section">
  <h3 class="case-study__heading">Technology</h3>
    <div class="case-study__text">

>       <p>Given the requirement of <strong>high performance, cross-platform code sharing</strong>, and the need for <strong>integration with their current software</strong>, Rust was a natural choice as a language with which to implement the architecture.<br>

Reviewer 2: I think "high performance" is not really the angle. It is also a bit vague—is it performant in terms of high throughput or low latency?
What we're really after here is _efficient_ processing.

Reviewer 2: To be more precise, we're after high availability without breaking the bank and with no risk of behaviour divergence.

>       <p>Given the requirements of <strong>efficiency, high availability, cross-platform code sharing</strong>, and the need for <strong>integration with their current software</strong>, Rust was a natural choice as a language with which to implement the design.<br>
>       What’s more, <a href="https://pola.rs/">Polars</a>, the famous high-performance open-source data frame manipulation library, is implemented in Rust and <a href="https://crates.io/crates/polars">published</a> on Rust’s package registry, allowing our engineers to <strong>trivially import it as a dependency</strong> in their application and reap the <strong>benefits of executing data processing in Rust</strong>.</p>
>       <p>Rust has proven to be an excellent basis for developing performant applications that are required to run on multiple platforms, including the end user’s browser.</p>

Reviewer 3: ```suggestion
      <p>Rust, with its strong performance and stability characteristics and solid WASM support, was a natural choice as a language with which to implement the design. What’s more, <a href="https://pola.rs/">Polars</a>, the famous high-performance open-source data frame manipulation library, is implemented in Rust itself and easily usable in Rust projects including those targeting WASM.</p>
```

    </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">Aftercare</h3>
    <div class="case-study__text">

>       <p>Upon finishing and delivering the proof-of-concept, Mainmatter provided guidance to their team on integrating it into KISTERS’ platform.</p>

Reviewer 3: ```suggestion
      <p>Upon finishing and delivering the proof-of-concept, Mainmatter provided guidance to KISTERS' team on integrating the code into their platform.</p>
```

---
layout: case-study
company: KISTERS
problem: KISTERS wanted to use Rust's cross-platform support to its full potential.
solution: We created a data processing package suitable for server, edge, and WASM.
tags: Launch your idea
title: Dynamically switching execution platforms with Rust | Work
displayTitle: "Dynamically switching execution platforms with Rust"
description: <p>KISTERS builds solutions for collecting, analyzing, and providing information based on environmental data.</p><p>The department we’ve been working with provides software that can help warn people of upcoming floods using data obtained from numerous sensor devices deployed throughout the globe.</p>
hero:
  color: purple
  image: "/assets/images/work/aleph-alpha-background-2.jpg" # TODO
  imageAlt: "TODO" # TODO
  tags: "development / architecture / Rust"
og:

>   image: /assets/images/cases/cs-aleph-alpha-og-image.jpg # TODO

Author: TODO

---

{% from "image-aspect-ratio.njk" import imageAspectRatio %} {% from "quote.njk" import quote %}

<div class="case-study__section">
  <h3 class="case-study__heading">About KISTERS</h3>
  <div class="case-study__text">
    <p>Based in Aachen, Germany, <a href="http://kisters.eu/">KISTERS</a> provides software for collecting, analyzing, and providing information based on environmental data.<br>
      The department we’ve been working with provides solutions that can help warn people of upcoming floods using data obtained from numerous sensor devices deployed throughout the globe.</p>
  </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">The challenge</h3>
    <div class="case-study__text">

>       <p>With big data comes big traffic. KISTERS wanted to <strong>handle large traffic spikes by dynamially switching</strong> between serving from a traditional server, from an edge function, or by offloading cumputation to end user's devices.</p>

Author: I'll change it to 'computation'

      <p><strong>To handle traffic spikes</strong>, KISTERS’ sought to employ a design that would allow them to <strong>quickly deploy their data processing to edge functions</strong> and, in case of extreme traffic surges, allow the processing steps to be <strong>run within the browser</strong> on end user’s devices.</p>
    </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">How Mainmatter helped out</h3>
    <div class="case-study__text">
        <p>During an initial development cycle, Luca Palmieri, Mainmatter’s Principal Engineering Consultant, and Founder and Engineer Marco Otte-Witte proposed and implemented an architecture that isolates the implementation of the data processing steps in a platform-independent package<p>
        <p>That way, the package can be thinly wrapped for use by a web server, served as an edge function, or compiled to Web Assembly for execution in browsers.</p>

>     </div>

Author: The next section mentions the requirement of high performance. I'd like the 'How Mainmatter helped out' to focus on describing at a high level what we did, and then 'technology' to detail the design choices made

</div>

<section class="mt-5">
        <!-->TODO add image for KISTERS<-->
        {% set imageData = {
          "imgPath": "/assets/images/work/aleph-alpha-graphs.jpg",

>           "alt": "TODO",

Author: TODO

      <p>Rust has proven to be an excellent basis for developing performant applications that are required to run on multiple platforms, including the end user’s browser.</p>
    </div>
</div>

<div class="case-study__section">
  <h3 class="case-study__heading">Aftercare</h3>
    <div class="case-study__text">
      <p>Upon finishing and delivering the proof-of-concept, Mainmatter provided guidance to their team on integrating it into KISTERS’ platform.</p>
      <p>Mainmatter has reviewed and optimized the integration code during several pairing sessions, ensuring that not just the product but also the knowledge required to maintain and expand it is transferred to KISTERS’ engineering team.</p>
    </div>  
</div>

<!-->TODO logo<-->
{% set content = {
"text": "TODO add quote",

> "source": "TODO add reference",

Author: TODO

"logo": "/assets/images/logos/monochrome/aleph-alpha.svg",
"logoAlt": "KISTERS logo",
"logoLoading": "lazy"

> } %} {{ quote(content) }}

Reviewer 3: I don't know whether we have a quote but we can also leave this out
